api/api.go

```diff
@@ -0,0 +1,387 @@
+	http     *http.Client
+	host     string
+	token    string
+	showHTTP bool
+}
+
+// ClientOptions are options for the API client.
+type ClientOptions struct {
+	// Insecure is a flag that indicates whether or not to supress SSL errors.
+	Insecure bool
+
+	// UseCerts is a flag that indicates whether system certs should be loaded
+	UseCerts bool
+
+	// Timeout specifies a time limit for requests made by this client.
+	Timeout time.Duration
+
+	// ShowHTTP is a flag that indicates whether or not HTTP requests and
+	// responses should be logged to stdout
+	ShowHTTP bool
+}
+
+// New returns a new API client.
+func New(
+	ctx context.Context,
+	host string,
+	opts *ClientOptions) (Client, error) {
+
+	if host == "" {
+		return nil, errNewClient
+	}
+
+	host = strings.Replace(host, "/api", "", 1)
+
+	c := &client{
+		http: &http.Client{},
+		host: host,
+	}
+
+	if opts != nil {
+		if opts.Timeout != 0 {
+			c.http.Timeout = opts.Timeout
+		}
+
+		if opts.Insecure {
+			c.http.Transport = &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			}
+		}
+
+		if opts.UseCerts {
+			pool, err := x509.SystemCertPool()
+			if err != nil {
+				return nil, errSysCerts
+			}
+			c.http.Transport = &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs:            pool,
+					InsecureSkipVerify: opts.Insecure,
+				},
+			}
+		}
+
+		if opts.ShowHTTP {
+			c.showHTTP = true
+		}
+	}
+
+	return c, nil
+}
+
+func (c *client) Get(
+	ctx context.Context,
+	path string,
+	headers map[string]string,
+	resp interface{}) error {
+
+	return c.DoWithHeaders(
+		ctx, http.MethodGet, path, headers, nil, resp)
+}
+
+func (c *client) Post(
+	ctx context.Context,
+	path string,
+	headers map[string]string,
+	body, resp interface{}) error {
+
+	return c.DoWithHeaders(
+		ctx, http.MethodPost, path, headers, body, resp)
+}
+
+func (c *client) Put(
+	ctx context.Context,
+	path string,
+	headers map[string]string,
+	body, resp interface{}) error {
+
+	return c.DoWithHeaders(
+		ctx, http.MethodPut, path, headers, body, resp)
+}
+
+func (c *client) Delete(
+	ctx context.Context,
+	path string,
+	headers map[string]string,
+	resp interface{}) error {
+
+	return c.DoWithHeaders(
+		ctx, http.MethodDelete, path, headers, nil, resp)
+}
+
+func (c *client) Do(
+	ctx context.Context,
+	method, path string,
+	body, resp interface{}) error {
+
+	return c.DoWithHeaders(ctx, method, path, nil, body, resp)
+}
+
+func beginsWithSlash(s string) bool {
+	return s[0] == '/'
+}
+
+func endsWithSlash(s string) bool {
+	return s[len(s)-1] == '/'
+}
+
+func (c *client) DoWithHeaders(
+	ctx context.Context,
+	method, uri string,
+	headers map[string]string,
+	body, resp interface{}) error {
+
+	res, err := c.DoAndGetResponseBody(
+		ctx, method, uri, headers, body)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// parse the response
+	switch {
+	case res == nil:
+		return nil
+	case res.StatusCode >= 200 && res.StatusCode <= 299:
```

Comment:
Hi @codenrhoden,

Please be aware of Go's [default HTTP redirect policy](https://golang.org/pkg/net/http/#Client). Without a custom redirect callback set, Go will stop trying after 10 attempts. You may wish to handle 3xx error codes separately in that case.

Reply:
@akutz thanks for making me aware. Unless you have a strong objection, I'm going to leave it for now. We weren't handling this in a separate way before. Also, if I understand the scenario correctly, if someone is using DNS to reach the API endpoint (rather than an IP), and they've set up >10 redirects, they might want to address that.